### PR TITLE
Add loongarch64 support

### PR DESCRIPTION
**Pull request by zhaixiaojuan**

### CODE REVIEW

File: include/ring-core/base.h [hunk 1]
```diff
@@ -83,6 +83,9 @@
 #elif defined(__ARMEL__) || defined(_M_ARM)
 #define OPENSSL_32_BIT
 #define OPENSSL_ARM
+#elif defined(__loongarch64)
```
Comment (xen0n) [inline include/ring-core/base.h:86]:
```suggestion
#elif defined(__loongarch_lp64)
```

File: mk/cargo.sh [hunk 2]
```diff
@@ -108,6 +108,11 @@ case $target in
       export CARGO_TARGET_X86_64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="$rustflags_self_contained"
     fi
     ;;
+  loongarch64-unknown-linux-gnuf64)
+    export CC_loongarch64_unknown_linux_gnuf64=clang-$llvm_version
+    export CC_loongarch64_unknown_linux_gnuf64=llvm-ar-$llvm_version
```
Comment (xen0n) [inline mk/cargo.sh:113]: These need to become `loongarch64-unknown-linux-gnu` after loongson/LoongArch-Documentation#80.

File: mk/install-build-tools.sh [hunk 3]
```diff
@@ -72,6 +72,9 @@ case $target in
 --target=i686-unknown-linux-musl|--target=x86_64-unknown-linux-musl)
   use_clang=1
   ;;
+--target=loongarch64-unknown-linux-gnuf64)
```
Comment (xen0n) [inline mk/install-build-tools.sh:75]: Also here.

**Review by xen0n — COMMENTED** [review]
The LLVM version needs to be bumped to 16 too before we can enable the build. Also, https://github.com/rust-lang/rust/pull/96971 is still not merged yet, let's at least wait for that so we can build *anything at all*.

**Comment by zhaixiaojuan** [comment]
RFR, Thanks~

File: mk/cargo.sh [hunk 4]
```diff
@@ -109,6 +109,11 @@ case $target in
       export CARGO_TARGET_X86_64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="$rustflags_self_contained"
     fi
     ;;
+  loongarch64-unknown-linux-gnu)
+    export CC_loongarch64_unknown_linux_gnu=clang-$llvm_version
+    export CC_loongarch64_unknown_linux_gnu=llvm-ar-$llvm_version
+    export CARGO_TARGET_LOONGARCH64_UNKNOWN_LINUX_GNUF64_LINKER=clang-$llvm_version
```
Comment (xen0n) [inline mk/cargo.sh:115]:
```suggestion
    export CARGO_TARGET_LOONGARCH64_UNKNOWN_LINUX_GNU_LINKER=clang-$llvm_version
```

**Review by xen0n — COMMENTED** [review]
Now the only remaining piece is the `$llvm_version` bump. Ideally that should be done for all arches, so you could do it in a separate PR.

File: mk/cargo.sh [hunk 5]
```diff
@@ -109,6 +109,11 @@ case $target in
       export CARGO_TARGET_X86_64_UNKNOWN_LINUX_MUSL_RUSTFLAGS="$rustflags_self_contained"
     fi
     ;;
+  loongarch64-unknown-linux-gnu)
+    export CC_loongarch64_unknown_linux_gnu=clang-$llvm_version
+    export CC_loongarch64_unknown_linux_gnu=llvm-ar-$llvm_version
```
Comment (aceice-x) [inline mk/cargo.sh:114]:
```suggestion
    export AR_loongarch64_unknown_linux_gnu=llvm-ar-$llvm_version
```

Comment (heiher) [reply mk/cargo.sh:114]: Thanks. I will create a new PR to do this since I have lost authentication to the current PR repository.

**Comment by heiher** [comment]
Tracked in new PR: https://github.com/briansmith/ring/pull/1632

**Comment by briansmith** [comment]
> Tracked in new PR: https://github.com/briansmith/ring/pull/1632

Thanks. I am closing this one.

**Pull request closed without merging**